### DISCUSSION TITLE
Help with moving an Item from Container A to Container B

### DISCUSSION THREAD

**Discussion by dmlo** [prisma/prisma#8077] [Q&A] [answered]
Hello Prisma people! Could you lend me a hand with a transaction problem?

With the schema 
```prisma
model Container {
  id              Int       @id @default(autoincrement())
  lockVersion     Int
  massLimit       Float
  items           Item[] 
}

model Item {
  id              Int       @id @default(autoincrement())
  type            String
  containers      Container @relation(fields: [containerId], references: [id])
  containerId     Int
  mass            Float
}
```

```Item``` is a real, physical object and thus can not exist in two ```Container```s at once. It must always be in a ```Container```.

I want to move an ```Item``` from ```Container``` A to ```Container``` B. The operation should fail if there is not enough remaining massLimit in ```Container``` B.

So the traditional solution here is 
1. Open a transaction
2. Check destination (B) massLimit
3. If ok, set item container to container B
4. if not ok, fail
5. Close transaction

In Prisma, this is considered a long-running transaction, which is (by design) not implemented. I understand the reasoning behind this and agree with it.

I am having difficulty implementing this item movement, however.

How should I set it up in Prisma?

My first thought was to attach a version field to the containers as suggested in https://www.prisma.io/docs/guides/performance-and-optimization/prisma-client-transactions-guide/#optimistic-concurrency-control .

```ts
    const sourceContainerId = 0;
    const destinationContainerId = 1;
    const itemId = 0;

    const container = await this.prisma.container.findUnique({
      where: {
        id: destinationContainerId,
      },
      include: {
        items: true,
      }
    });

    const item = await this.prisma.item.findUnique({where: {id: itemId}});

    const massRemaining = container.massLimit - massOfItemsinContainer( container );
    if (item.mass > massRemaining) { return false; } 

    const updated = await this.prisma.container.updateMany({
      where:{
        id: destinationContainer.id,
        lockVersion: container.lockVersion
      }, 
      data:{
        items:{
          connect:{
            id: item.id
          }
        }
      }
    });

    if (updated.count < 1) { return false; }
    return true;
```

This does not work because Prisma allows me to access direct properties of the Container table in the nested operation but not the relation:

```
Type '{ items: { connect: { id: number; }; }; }' is not assignable to type
 '(Without<ContainerUpdateManyMutationInput, ContainerUncheckedUpdateManyInput> & 
ContainerUncheckedUpdateManyInput) | (Without<...> & ContainerUpdateManyMutationInput)'.

  Object literal may only specify known properties, and 'items' does not exist in type 
'(Without<ContainerUpdateManyMutationInput, ContainerUncheckedUpdateManyInput> & 
ContainerUncheckedUpdateManyInput) | (Without<...> & ContainerUpdateManyMutationInput)'.ts(2322)
```

If I set it up in the other direction, accessing through Item to Container, I don't see a way to ensure that container B has not been modified since I checked its massLimit remaining.

Is there a Prisma way to solve this problem without resorting to a raw SQL transaction?

Any advice is much appreciated! Thanks for helping me out.

**Comment by ryands17** [comment, marked as answer]
@dmlo 👋 

I can't find any other way to do this apart from a [raw SQL transaction](https://www.prisma.io/docs/concepts/components/prisma-client/raw-database-access). We will be adding support for long running transactions so you can follow this request for more updates: https://github.com/prisma/prisma/issues/1844#issuecomment-846477520

**Reply by dmlo** [reply]
Thank you for taking a look at it.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
